### PR TITLE
Dorks Manager Test

### PR DESCRIPTION
**Pull request by rjt-gupta**
All the tests are working after #330 gets merged.

But, I need help on `choose_dorks` method as while running the test `random.sample` sometimes tries to sample more items from the list of lesser size. Hence throws `ValueError: sample larger than population`.

Please have a look at the possible workaround [here](https://stackoverflow.com/questions/20861497/sample-larger-than-population-in-random-sample-python).

### CODE REVIEW

**Comment by rnehra01** [comment]
I think you can mock `random.randint`.

**Comment by rjt-gupta** [comment]
I don't understand why test are failing here, it works fine locally.

**Comment by rnehra01** [comment]
I think you should check whether values are inserted correctly in redis instead of mocking library methods.

**Comment by rjt-gupta** [comment]
Done.

File: tanner/tests/test_dorks_manager.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    def test_push_init_dorks(self):
+
+        async def test():
+            await self.handler.push_init_dorks(config.TannerConfig.get('DATA', 'dorks'), self.handler.dorks_key,
+                                               self.redis_client)
+            self.returned_result = await self.redis_client.smembers(self.handler.dorks_key)
+
+        self.loop.run_until_complete(test())
+        self.assertEqual(len(self.returned_result), 3562)
```
Comment (rnehra01) [inline tanner/tests/test_dorks_manager.py]: Is `self.handler.dorks_key` == `3562`?
Then I guess you should substitute `self.handler.dorks_key` with some constant and check it here other wise tests will fail if there is some modification in `self.handler.dorks_key`

File: tanner/tests/test_dorks_manager.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        self.redis_client.sadd = AsyncMock(side_effect=aioredis.ProtocolError)
+
+        async def test():
+            await self.handler.extract_path(self.path, self.redis_client)
+
+        with self.assertLogs(level='ERROR') as log:
+            self.loop.run_until_complete(test())
+            self.assertIn('Problem with redis connection', log.output[0])
+
+    def test_init_dorks(self):
```
Comment (rnehra01) [inline tanner/tests/test_dorks_manager.py]: Check the case if any key doesn't exist
https://github.com/mushorg/tanner/blob/master/tanner/dorks_manager.py#L48
https://github.com/mushorg/tanner/blob/master/tanner/dorks_manager.py#L49

File: tanner/tests/test_dorks_manager.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        ]
+
+        async def test():
+            await self.handler.init_dorks(self.redis_client)
+
+        self.loop.run_until_complete(test())
+        self.handler.push_init_dorks.assert_has_calls(calls)
+
+    def test_choose_dorks(self):
+        random.randint = mock.Mock(return_value=0)
```
Comment (rnehra01) [inline tanner/tests/test_dorks_manager.py]: Please do better testing. Test return value.

File: .travis.yml [hunk 4]
```diff
@@ -17,6 +17,7 @@ install:
   - docker-compose -f ./docker/docker-compose.yml up -d
 # command to run tests
 script:
+    - python setup.py install
```
Comment (rnehra01) [inline .travis.yml]: How does this help?

File: tanner/tests/test_dorks_manager.py (hunk 1 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: No, `self.handler.dorks_key` is not equal to 3562. 3562 is the total length of the list which is returned by `smembers`. Here `self.handler.dorks_key` == `9445102ddf663ed0a20382c219a952ba`

File: tanner/tests/test_dorks_manager.py (hunk 3 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: Here its the problem if I choose values greater than zero it throws the error `ValueError: sample larger than population`. I think we don't have to test randomly chosen dorks.

File: .travis.yml (hunk 4 above)
Comment (rjt-gupta) [reply .travis.yml]: The problem was there is no `dorks.pickle` and `user_dorks.pickle` file which is created during `setup`.

 https://github.com/mushorg/tanner/blob/9fa7289896d194e0ca7b03faec3dfd72553271a8/tanner/dorks_manager.py#L59

File: tanner/tests/test_dorks_manager.py (hunk 1 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: I guess total length should remain same.

File: tanner/tests/test_dorks_manager.py (hunk 2 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: Done.

Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: No I meant to add a case when one or both have None values.

File: tanner/tests/test_dorks_manager.py (hunk 3 above)
Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: I guess you need to dig deeper. Above you're saying the number is near 3k and here this error, so please look into it.

File: tanner/tests/test_dorks_manager.py (hunk 1 above)
Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: As far I understand this is loaded from the pickle which can change.

File: tanner/tests/test_dorks_manager.py (hunk 2 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: Got it.

File: tanner/tests/test_dorks_manager.py (hunk 3 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: Both cases are different here the problem is with `random.sample` not dorks keys.

File: tanner/tests/test_dorks_manager.py [hunk 5]
```diff
@@ -25,14 +25,14 @@ def setUp(self):
...
+        self.redis_client.sadd = AsyncMock()
 
         async def test():
             await self.handler.push_init_dorks(config.TannerConfig.get('DATA', 'dorks'), self.handler.dorks_key,
                                                self.redis_client)
-            self.returned_result = await self.redis_client.smembers(self.handler.dorks_key)
 
         self.loop.run_until_complete(test())
-        self.assertEqual(len(self.returned_result), 3562)
+        assert self.redis_client.sadd.called
```
Comment (rnehra01) [inline tanner/tests/test_dorks_manager.py]: Don't mock library methods.

File: tanner/tests/test_dorks_manager.py (hunk 3 above)
Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: As per the errors you got, can you the actual population/size?

File: tanner/tests/test_dorks_manager.py (hunk 5 above)
Comment (rjt-gupta) [reply tanner/tests/test_dorks_manager.py]: I used this to just convert it into a asyncmock object to test if `sadd` is called.

Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: I understand that but I prefer the previous version of tests.

Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: I understood that but I prefer the previous version.

Comment (rnehra01) [reply tanner/tests/test_dorks_manager.py]: Can you try passing a custom pickle with a few data samples?

**Comment by rjt-gupta** [comment]
Updated both methods using custom pickle file. 😀

File: .travis.yml (hunk 4 above)
Comment (rnehra01) [reply .travis.yml]: Is this still required?

Comment (rjt-gupta) [reply .travis.yml]: For now its not required, but I think its good to have this.

**Comment by rnehra01** [comment]
I don't think it's required to make unnecessary changes.

**Comment by rjt-gupta** [comment]
> 
> I don't think it's required to make unnecessary changes.

Okay Removed.

**Pull request merged**